internal/http/services/appprovider/appprovider.go

```diff
@@ -0,0 +1,232 @@
+	}
+	if c.AccessTokenTTL == 0 {
+		c.AccessTokenTTL = 86400
+	}
+	c.GatewaySvc = sharedconf.GetGatewaySVC(c.GatewaySvc)
+}
+
+type svc struct {
+	conf *Config
+}
+
+// New returns a new ocmd object
+func New(m map[string]interface{}, log *zerolog.Logger) (global.Service, error) {
+
+	conf := &Config{}
+	if err := mapstructure.Decode(m, conf); err != nil {
+		return nil, err
+	}
+	conf.init()
+
+	s := &svc{
+		conf: conf,
+	}
+	return s, nil
+}
+
+// Close performs cleanup.
+func (s *svc) Close() error {
+	return nil
+}
+
+func (s *svc) Prefix() string {
+	return s.conf.Prefix
+}
+
+func (s *svc) Unprotected() []string {
+	return []string{}
+}
+
+func (s *svc) Handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			ocmd.WriteError(w, r, ocmd.APIErrorUnimplemented, "only GET requests are supported", errors.New("only GET requests are supported"))
+			return
+		}
+
+		s.handleWopiOpen(w, r)
+	})
+}
+
+// WopiResponse holds the various fields to be returned for a wopi open call
+type WopiResponse struct {
+	WopiClientURL  string `json:"wopiclienturl"`
+	AccessToken    string `json:"accesstoken"`
+	AccessTokenTTL int64  `json:"accesstokenttl"`
+}
+
+func (s *svc) handleWopiOpen(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	client, err := pool.GetGatewayServiceClient(s.conf.GatewaySvc)
+	if err != nil {
+		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error getting grpc gateway client", err)
+		return
+	}
+
+	info, errCode, err := s.getStatInfo(ctx, r.URL.Query().Get("fileId"), client)
+	if err != nil {
+		ocmd.WriteError(w, r, errCode, "error statting file", err)
+	}
+
+	openReq := gateway.OpenInAppRequest{
+		Ref:      &provider.Reference{ResourceId: info.Id},
+		ViewMode: getViewMode(info),
+		App:      r.URL.Query().Get("app"),
+	}
+	openRes, err := client.OpenInApp(ctx, &openReq)
+	if err != nil {
+		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error opening resource", err)
+		return
+	}
+	if openRes.Status.Code != rpc.Code_CODE_OK {
+		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error opening resource information", status.NewErrorFromCode(openRes.Status.Code, "appprovider"))
+		return
+	}
+
+	u, err := url.Parse(openRes.AppUrl)
+	if err != nil {
+		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error parsing app URL", err)
+		return
+	}
+	q := u.Query()
+
+	// remove access token from query parameters
+	accessToken := q.Get("access_token")
+	q.Del("access_token")
+
+	// more options used by oC 10:
+	// &lang=en-GB
+	// &closebutton=1
+	// &revisionhistory=1
+	// &title=Hello.odt
+	u.RawQuery = q.Encode()
+
+	js, err := json.Marshal(
+		WopiResponse{
+			WopiClientURL: u.String(),
+			AccessToken:   accessToken,
+			// https://wopi.readthedocs.io/projects/wopirest/en/latest/concepts.html#term-access-token-ttl
+			AccessTokenTTL: time.Now().Add(time.Second*time.Duration(s.conf.AccessTokenTTL)).UnixNano() / 1e6,
```

Comment:
(Related to the above) this configuration parameter is specified in the WOPI server config, and would need to be duplicated in some Reva config. Why is it important to provide this information to the web UI?

Reply:
We are currently using this in the oCIS WOPI extension to not have the access token in the browsers' address bar. If you have the access token in the address bar it could easily leak to other people (copy and paste)

https://github.com/owncloud/ocis-wopiserver/blob/e23ff205ec4474e4297ba39248636dca5fbb9402/ui/app.js#L84-L105 shows the form push with the access token. Downside is that you cant copy the URL and open it in a new tab

Reply:
The same form post method is also recommended for embedding into iframes: https://wopi.readthedocs.io/en/latest/hostpage.html

Reply:
Makes sense, indeed we currently embed those URLs in an iframe so the user only sees a cernbox URL and can't really copy/paste it (yet of course by inspecting the iframe's content you'd see the access_token).

At this point I'd at least rename the struct to `AppResponse`, including the fields `AppURL` and `AccessToken`, just to be more generic and not WOPI specific. Other applications that do not implement an access token mechanism may leave it empty. Also, I'd still drop the TTL value - eventually, what rules is the TTL of the main JWT obtained by the user when logging in the first time, right?

Reply:
Depending on how you define a TTL, the access token ttl is WOPI specific since it is the UNIX timestamp of the moment when the storage / REVA JWT will expire.
In Collabora providing access_token_ttl information triggers a warning for the user that editing will be no longer possible after the expiry (because after that Collabora cannot save the document obviously). 

https://wopi.readthedocs.io/projects/wopirest/en/latest/concepts.html#term-access-token-ttl:
```
To prevent data loss for users, Office for the web will prompt users to save and refresh their sessions if the access token for their session is close to expiring. In order to do this, Office for the web needs to know when the access token will expire, which it determines based on the access_token_ttl value.
```

Reply:
ok, to be exact it's the `min(Reva_JWT_TTL, WOPI_access_token_TTL)`. I must say I had forgotten about this feature of the WOPI protocol, but then the value must be computed on the fly, not taken from the config (yeah, as admitted by the WOPI docs, the name is misleading, it should really be called `access_token_expiration_time`): can we extract the expiration time of a Reva JWT? If so, we can safely assume that the WOPI access_token's validity is never shorter (its default value in the WOPI server config is `24h`) and return the Reva JWT TTL.

Reply:
yeah of course we can take the `exp` property from the WOPI jwt access token (if we do so without validating the WOPI access token, we also don't need the WOPI jwt secret in here, which is fine for only taking the `exp`  from it)

Reply:
If we have the expiration time of the REVA jwt access token, that's more than enough - we can assume the WOPI jwt access token is issued with a _longer_ expiration in all cases - at least it should be configured like that.